velox/docs/develop/window.rst

```diff
@@ -205,3 +205,107 @@ peer row with a NULL and the frame_end index is the last peer row with a NULL va
 
 Rows with NULL values do not participate in the frames of the other rows.
 
+
+Empty frames
+------------
+
+Its possible that certain rows have empty frames. Empty frames occur when :
+
+* Both frame_start and frame_end fall before the first partition row.
+
+  E.g. in frame *ROWS BETWEEN 5 PRECEDING and 2 PRECEDING* the first 2 rows
+  have both frame bounds before the first partition row.
+
+* Both frame_start and frame_end fall after the partition end row.
+
+  E.g. in frame *ROWS BETWEEN 2 FOLLOWING and 5 FOLLOWING* the last 2 rows
+  have both frame bounds beyond the last partition row.
+
+* frame_start > frame_end row (as frame range is defined from frame_start to
+  frame_end).
+
+  E.g. In frame *ROWS BETWEEN UNBOUNDED PRECEDING AND 2 PRECEDING* the intent
+  is to compute aggregation from the partition start row to 2 rows prior
+  the current one. However, for the first 2 rows the frameStart
+  (frame index 0 for unbounded preceding) is ahead of 2 preceding
+  (indices -2 and -1).
+
+* For frames like *ROWS BETWEEN 2 PRECEDING AND 5 PRECEDING* or
+  *ROWS BETWEEN 5 FOLLOWING AND 2 FOLLOWING*, frame_start > frame_end for
+  all rows. So all frames are empty.
+
+**Partial frames**
+
+Like illustrated in the examples above, rows could have partial window frames.
+
+A partial frame occurs when:
+
+- frame_start < frame_end (so the empty frame condition is not violated)
+- One frame end is within partition bounds and the other end outside of it.
+  In these cases either frame_start is before the first partition row while
+  frame_end within the partition, or frame_start is in the partition but
+  frame_end beyond it.
+
+For such frames the violating boundary is clamped to the first or last row
+of the partition.
```

Comment:
Could we clarify this a bit by discussing the two cases separately? E.g., frame_start before the partition bound should be clamped to the first row of the partition. frame_end after the partition bound should be clamped to the last row of the partition.

Reply:
Done.